### PR TITLE
Xml Writer

### PR DESCRIPTION
**Pull request by angela2389**
Hi Microsoft team,

Can you please review this Pull Request? Hopefully this will be the last one :)

### CODE REVIEW

**Comment by JesperSchulz** [comment]
Before we start looking into your PR, could you elaborate on where our [native XML Document implementation](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/xmldocument/xmldocument-data-type) doesn't suffice? Instead of creating a competing XML writer, could we add to our native implementation to suite your needs?

**Comment by angela2389** [comment]
Hi @JesperSchulz,

I do use XML Documents and it works well for small files, but in case of big XML files we have a lot of performance issues. Customers get impatient if the loading time goes over 1 minute. Using Xml Writer vs Xml Document makes a huge difference.

Please see the results of @AndersMad (we are not in the same company) on https://github.com/microsoft/AL/issues/999 and Business Central Idea: https://experience.dynamics.com/ideas/idea/?ideaid=f07b7746-6c80-e911-80e7-0003ff68bdb5

**Comment by JesperSchulz** [comment]
Got it, thanks! Let's see what we can do here :-)

File: Modules/System/XmlWriter/src/XmlWriterImpl.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    end;
+
+    procedure XmlWriterElementString(LocalName: Text; ElementValue: Text)
+    begin
+        XmlTextWriter.WriteElementString(LocalName, ElementValue);
+    end;
+
+    procedure XmlWriterEndElement()
+    begin
+        XmlTextWriter.WriteEndElement;
```
Comment (JesperSchulz) [inline Modules/System/XmlWriter/src/XmlWriterImpl.Codeunit.al:36]: Let's keep the original function names. If the method was called WriteEndElement, let's call our AL function WriteEndElement.

File: Modules/System/XmlWriter/src/XmlWriterImpl.Codeunit.al [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    end;
+
+    procedure XmlWriterAddAttribute(ElementName: Text; ElementValue: Text; Prefix: Text)
+    begin
+        if Prefix <> '' then
```
Comment (JesperSchulz) [inline Modules/System/XmlWriter/src/XmlWriterImpl.Codeunit.al:41]: Let's not put logic, where none is needed. Remember, that we also have function-overloads in al now!

File: Modules/System/XmlWriter/src/XmlWriter.Codeunit.al [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+/// <summary>
+// Provides helper functions for System.Xml.XmlWriter
+/// </summary>
+
+codeunit 1385 "XmlWriter"
+{
+    Access = Public;
+
+    var
+        XmlWriterImpl: Codeunit "XmlWriter Impl.";
```
Comment (JesperSchulz) [inline Modules/System/XmlWriter/src/XmlWriter.Codeunit.al:15]: I wonder, if we even need an Impl. codeunit for this one? Right now, the facade calls the impl. codeunit with a one-liner. The impl. codeunit the again calls the .NET method with another one-liner - seems a little over-wrapped to me :-)

Comment (angela2389) [reply Modules/System/XmlWriter/src/XmlWriter.Codeunit.al:15]: If it's fine for you to do without I will remove it :), I'm just trying to follow the template of the other modules.

Comment (angela2389) [reply Modules/System/XmlWriter/src/XmlWriter.Codeunit.al:15]: Actually, the XmlWriterCreateDocument function has four lines now..

**Comment by JesperSchulz** [comment]
We will process your request within the month of September :-)

**Review by JesperSchulz — APPROVED** [review]

**Comment by bc-ghost** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. We will update this issue with information about availability.

**Comment by bc-ghost** [comment]
Thanks for submitting this issue. We will publish a fix for it in the next minor update for release 17.

**Comment by bc-ghost** [comment]
Thanks for submitting this issue. We will publish a fix for it in the next minor update for release 16.

**Pull request merged**